File: windows/TitleBarColorizer.cs
using System.IO;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;

namespace CursorProfiles;

/// <summary>
/// A running Cursor window's taskbar icon can't be rebranded (re-signing
/// Cursor's own executable under a different identity crashes it — this is
/// deliberate anti-tamper protection, confirmed against a real install), so
/// the next best way to tell profiles apart while they're open is a distinct
/// title bar color per profile, visible in the window itself and Alt+Tab.
/// </summary>
public static class TitleBarColorizer
{
    /// <summary>Merge this profile's title-bar colors into its
    /// User/settings.json, preserving every other setting already there.
    /// No-op for the built-in profile — the user's real settings are never touched.</summary>
    public static void Apply(CursorProfile profile, string profileDir)
    {
        if (profile.IsSystem) return;

        var userDir = Path.Combine(profileDir, "User");
        Directory.CreateDirectory(userDir);
        var settingsPath = Path.Combine(userDir, "settings.json");

        var settings = ReadSettings(settingsPath);

        var colors = settings["workbench.colorCustomizations"] as JsonObject ?? new JsonObject();
        foreach (var (key, value) in ColorValues(profile.ColorHex))
            colors[key] = value;
        settings["workbench.colorCustomizations"] = colors;
        // The default title bar ignores colorCustomizations — required for the colors to render.
        settings["window.titleBarStyle"] = "custom";

        File.WriteAllText(settingsPath, settings.ToJsonString(new JsonSerializerOptions { WriteIndented = true }));
    }

    public static Dictionary<string, string> ColorValues(string hex)
    {
        var foreground = ContrastingForeground(hex);
        return new()
        {
            ["titleBar.activeBackground"] = hex,
            ["titleBar.activeForeground"] = foreground,
            ["titleBar.inactiveBackground"] = Shade(hex, 0.55),
            ["titleBar.inactiveForeground"] = foreground + "AA",
        };
    }

    // MARK: JSONC-tolerant read

    private static JsonObject ReadSettings(string path)
    {
        if (!File.Exists(path)) return new JsonObject();
        try
        {
            var stripped = StripJsonComments(File.ReadAllText(path));
            return JsonNode.Parse(stripped) as JsonObject ?? new JsonObject();
        }
        catch (JsonException)
        {
            return new JsonObject(); // unparsable settings.json — start fresh rather than fail the launch
        }
    }

    /// <summary>Strips // and /* */ comments and trailing commas so a
    /// hand-edited JSONC settings.json (VS Code allows both) round-trips
    /// through strict JSON parsing. Comments are not preserved on write.</summary>
    private static string StripJsonComments(string text)
    {
        var sb = new System.Text.StringBuilder(text.Length);
        bool inString = false, escaped = false;
        for (int i = 0; i < text.Length; i++)
        {
            char c = text[i];
            if (inString)
            {
                sb.Append(c);
                if (escaped) escaped = false;
                else if (c == '\\') escaped = true;
                else if (c == '"') inString = false;
                continue;
            }
            if (c == '"') { inString = true; sb.Append(c); continue; }
            if (c == '/' && i + 1 < text.Length)
            {
                if (text[i + 1] == '/')
                {
                    while (i < text.Length && text[i] != '\n') i++;
                    i--;
                    continue;
                }
                if (text[i + 1] == '*')
                {
                    i += 2;
                    while (i + 1 < text.Length && !(text[i] == '*' && text[i + 1] == '/')) i++;
                    i++;
                    continue;
                }
            }
            sb.Append(c);
        }
        return Regex.Replace(sb.ToString(), @",\s*([}\]])", "$1");
    }

    // MARK: Color helpers

    private static string Shade(string hex, double factor)
    {
        if (!uint.TryParse(hex.TrimStart('#'), System.Globalization.NumberStyles.HexNumber, null, out var value))
            return hex;
        var r = (int)(((value >> 16) & 0xFF) * factor);
        var g = (int)(((value >> 8) & 0xFF) * factor);
        var b = (int)((value & 0xFF) * factor);
        return $"#{r:X2}{g:X2}{b:X2}";
    }

    private static string ContrastingForeground(string hex)
    {
        if (!uint.TryParse(hex.TrimStart('#'), System.Globalization.NumberStyles.HexNumber, null, out var value))
            return "#FFFFFF";
        double r = (value >> 16) & 0xFF, g = (value >> 8) & 0xFF, b = value & 0xFF;
        var luminance = 0.299 * r + 0.587 * g + 0.114 * b;
        return luminance > 150 ? "#1A1A1A" : "#FFFFFF";
    }
}

File: Sources/CursorProfiles/TitleBarColorizer.swift
import AppKit

/// Cursor's Dock icon can't be rebranded per profile — re-signing its
/// executable under a different identity crashes it (deliberate anti-tamper
/// protection), which is also why this app no longer builds per-profile
/// wrapper .apps at all: each one is a distinct bundle identity, and macOS
/// ties permission consent (mic, camera, folder access, etc.) to whichever
/// bundle launched Cursor — so every wrapper re-triggered Cursor's entire
/// permission set as if it were a brand new, unfamiliar app. Instead, the
/// next best way to tell profiles apart while they're open is a distinct
/// title bar color per profile, visible in the window itself and in
/// Cmd+Tab/Mission Control.
enum TitleBarColorizer {

    /// Merge this profile's title-bar colors into its `User/settings.json`,
    /// preserving every other setting already there. No-op for the built-in
    /// profile — we never touch the user's real Cursor settings.
    static func apply(profile: CursorProfile, profileDir: URL) {
        guard !profile.isSystem else { return }

        let settingsURL = profileDir.appendingPathComponent("User/settings.json")
        do {
            try FileManager.default.createDirectory(
                at: settingsURL.deletingLastPathComponent(), withIntermediateDirectories: true)
        } catch {
            return
        }

        var settings = readSettings(at: settingsURL)

        var colors = settings["workbench.colorCustomizations"] as? [String: Any] ?? [:]
        for (key, value) in colorValues(for: profile.colorHex) {
            colors[key] = value
        }
        settings["workbench.colorCustomizations"] = colors
        // Native title bars ignore colorCustomizations on macOS — required for the colors to render.
        settings["window.titleBarStyle"] = "custom"

        writeSettings(settings, to: settingsURL)
    }

    static func colorValues(for hex: String) -> [String: String] {
        let foreground = contrastingForeground(for: hex)
        return [
            "titleBar.activeBackground": hex,
            "titleBar.activeForeground": foreground,
            "titleBar.inactiveBackground": shade(hex, by: 0.55),
            "titleBar.inactiveForeground": foreground + "AA",
        ]
    }

    // MARK: - JSONC-tolerant read/write

    private static func readSettings(at url: URL) -> [String: Any] {
        guard let data = try? Data(contentsOf: url),
              let text = String(data: data, encoding: .utf8) else { return [:] }
        let stripped = stripJSONComments(text)
        guard let strippedData = stripped.data(using: .utf8),
              let object = try? JSONSerialization.jsonObject(with: strippedData),
              let dict = object as? [String: Any] else { return [:] }
        return dict
    }

    private static func writeSettings(_ settings: [String: Any], to url: URL) {
        guard let data = try? JSONSerialization.data(
            withJSONObject: settings, options: [.prettyPrinted, .sortedKeys]) else { return }
        try? data.write(to: url, options: .atomic)
    }

    /// Strips `//` and `/* */` comments and trailing commas so a hand-edited
    /// JSONC settings.json (VS Code allows both) can still round-trip through
    /// strict JSONSerialization. Comments are not preserved on write.
    private static func stripJSONComments(_ text: String) -> String {
        var result = ""
        result.reserveCapacity(text.count)
        var inString = false
        var escaped = false
        var i = text.startIndex
        while i < text.endIndex {
            let c = text[i]
            if inString {
                result.append(c)
                if escaped { escaped = false }
                else if c == "\\" { escaped = true }
                else if c == "\"" { inString = false }
                i = text.index(after: i)
                continue
            }
            if c == "\"" {
                inString = true
                result.append(c)
                i = text.index(after: i)
                continue
            }
            if c == "/", text.index(after: i) < text.endIndex {
                let next = text[text.index(after: i)]
                if next == "/" {
                    while i < text.endIndex, text[i] != "\n" { i = text.index(after: i) }
                    continue
                }
                if next == "*" {
                    i = text.index(i, offsetBy: 2)
                    while i < text.endIndex {
                        if text[i] == "*", text.index(after: i) < text.endIndex,
                           text[text.index(after: i)] == "/" {
                            i = text.index(i, offsetBy: 2)
                            break
                        }
                        i = text.index(after: i)
                    }
                    continue
                }
            }
            result.append(c)
            i = text.index(after: i)
        }
        // Trailing commas before a closing bracket/brace.
        while let range = result.range(of: #",\s*([}\]])"#, options: .regularExpression) {
            let closer = result[result.index(before: range.upperBound)]
            result.replaceSubrange(range, with: String(closer))
        }
        return result
    }

    // MARK: - Color helpers

    private static func shade(_ hex: String, by factor: Double) -> String {
        guard let value = UInt32(hex.replacingOccurrences(of: "#", with: ""), radix: 16) else { return hex }
        let r = Double((value >> 16) & 0xFF) * factor
        let g = Double((value >> 8) & 0xFF) * factor
        let b = Double(value & 0xFF) * factor
        return String(format: "#%02X%02X%02X", Int(r), Int(g), Int(b))
    }

    private static func contrastingForeground(for hex: String) -> String {
        guard let value = UInt32(hex.replacingOccurrences(of: "#", with: ""), radix: 16) else { return "#FFFFFF" }
        let r = Double((value >> 16) & 0xFF)
        let g = Double((value >> 8) & 0xFF)
        let b = Double(value & 0xFF)
        let luminance = 0.299 * r + 0.587 * g + 0.114 * b
        return luminance > 150 ? "#1A1A1A" : "#FFFFFF"
    }
}
